File: integrations/langgraph/examples/02_agent_with_memory.py
"""A LangGraph agent that remembers across conversations.

This wires up all three capabilities at once:

* `store=`            — Cortadel as the agent's long-term memory backend (`BaseStore`)
* `tools=`            — `search_memory` / `add_memories`, for when the agent decides to look
* `pre/post_model_hook` — automatic recall and persistence, for when it doesn't

The point of the demo is the *second* run: it uses a different `thread_id`, so the checkpointer
gives it an empty transcript, and anything the agent still knows came from Cortadel.

Run it:

    docker compose up                      # from the cortadel repo root -> http://localhost:3001
    export ANTHROPIC_API_KEY=...           # or swap the model line for any LangChain chat model
    uv run --with cortadel-langgraph --with "langchain[anthropic]" python examples/02_agent_with_memory.py
"""
from __future__ import annotations

import os

from langchain_core.messages import HumanMessage
from langgraph.checkpoint.memory import InMemorySaver
from langgraph.prebuilt import create_react_agent

from cortadel_langgraph import CortadelMemory, CortadelStore, create_memory_tools

BASE_URL = os.environ.get("CORTADEL_URL", "http://localhost:3001")
API_KEY = os.environ.get("CORTADEL_API_KEY")

USER_ID = "e2e-langgraph-agent-demo"  # e2e-* — this writes to a real server
NAMESPACE = ("memories", USER_ID)

store = CortadelStore(BASE_URL, USER_ID, api_key=API_KEY)

# Automatic memory. `recall_node` searches Cortadel with the user's message and prepends the hits
# for this model call only (via `llm_input_messages`, which create_react_agent treats as
# ephemeral), so recalled memories never pile up in the transcript. `remember_node` hands each
# finished turn to Cortadel's conversation-ingest pipeline, which distils the durable facts.
memory = CortadelMemory(store, namespace=NAMESPACE)

agent = create_react_agent(
    model="anthropic:claude-sonnet-4-5",
    # The explicit tools are complementary, not redundant: the hooks cover "what is relevant to
    # what was just said", the tools cover "go and look something specific up / write this down".
    tools=create_memory_tools(store=store, namespace=NAMESPACE),
    prompt="You are a helpful assistant with a long memory. Be concise.",
    store=store,
    pre_model_hook=memory.pre_model_hook,
    post_model_hook=memory.post_model_hook,
    checkpointer=InMemorySaver(),
)


def say(text: str, thread_id: str) -> None:
    print(f"\n[thread {thread_id}] you: {text}")
    result = agent.invoke(
        {"messages": [HumanMessage(text)]},
        config={"configurable": {"thread_id": thread_id}},
    )
    print(f"[thread {thread_id}] agent: {result['messages'][-1].content}")


def main() -> None:
    if store.health(NAMESPACE) is None:
        raise SystemExit(f"No Cortadel server at {BASE_URL}. Start one with `docker compose up`.")

    # --- Conversation one -------------------------------------------------------------------
    say("Hi! I'm Robin. I only ship releases on Fridays, and I'm allergic to shellfish.", "conv-1")
    say("What's a good rule of thumb for release notes?", "conv-1")

    # Cortadel's extraction pipeline runs off the request path, so give it a moment before the
    # facts from conversation one become searchable.
    print("\n(waiting for Cortadel to extract facts from conversation one...)")
    import time

    time.sleep(5)

    # --- Conversation two: a brand-new thread, so the transcript is empty -------------------
    # Anything the agent knows here came out of Cortadel, not out of the checkpointer.
    say("When should we schedule the next deploy? And where should we get lunch?", "conv-2")


if __name__ == "__main__":
    try:
        main()
    finally:
        store.close()

File: integrations/langgraph/examples/02-agent-with-memory.ts
/**
 * A `createReactAgent` that remembers — automatic recall before the model, automatic persistence
 * after the turn, plus the two memory tools for when the agent wants to look something up itself.
 *
 * It deliberately runs on `FakeListChatModel` so the example needs **no LLM API key**: what is
 * being demonstrated is the memory wiring, and a canned model makes the effect visible without
 * spending anything. Swap in a real model with one line — see `REAL MODEL` below.
 *
 * Run it with a Cortadel server on :3001 (`docker compose up` from the repo root):
 *
 *   pnpm dlx tsx examples/02-agent-with-memory.ts
 */
import process from "node:process";
import { HumanMessage } from "@langchain/core/messages";
import { FakeListChatModel } from "@langchain/core/utils/testing";
import { MemorySaver } from "@langchain/langgraph";
import { createReactAgent } from "@langchain/langgraph/prebuilt";

import { CortadelMemory, CortadelStore, createMemoryTools } from "@cortadel/langgraph";

const BASE_URL = process.env.CORTADEL_BASE_URL ?? "http://localhost:3001";
const USER_ID = "e2e-langgraph-agent";
const NAMESPACE = ["memories", USER_ID];

async function main(): Promise<void> {
  // 1. Cortadel as the graph's long-term store.
  const store = new CortadelStore({
    baseUrl: BASE_URL,
    userId: USER_ID,
    apiKey: process.env.CORTADEL_API_KEY,
  });

  // 2. Automatic memory: recall before the model call, persist after the turn. Neither needs the
  //    agent to decide anything.
  const memory = new CortadelMemory({ store, namespace: NAMESPACE });

  // REAL MODEL: replace this with, say,
  //   import { ChatAnthropic } from "@langchain/anthropic";
  //   const llm = new ChatAnthropic({ model: "claude-sonnet-4-5" });
  const llm = new FakeListChatModel({
    responses: ["Noted — Fridays it is.", "Friday, going by what you told me before."],
  });

  const agent = createReactAgent({
    llm,
    // 3. ...and tools, for when the agent wants to look something up or write something down.
    tools: createMemoryTools({ store, namespace: NAMESPACE }),
    store,
    preModelHook: memory.preModelHook,
    postModelHook: memory.postModelHook,
    checkpointer: new MemorySaver(),
  });

  // Turn one, on thread `conv-1`.
  await agent.invoke(
    { messages: [new HumanMessage("I ship releases on Fridays.")] },
    { configurable: { thread_id: "conv-1" } },
  );

  // Turn two, on a brand-new thread — the checkpointer's transcript is empty here, so anything the
  // agent knows now came out of Cortadel.
  const result = await agent.invoke(
    { messages: [new HumanMessage("When should we deploy?")] },
    { configurable: { thread_id: "conv-2" } },
  );
  console.log(String(result.messages[result.messages.length - 1]?.content));

  // Long-term memory crossed the thread boundary; the transcript did not.
  const remembered = await store.search(NAMESPACE, { query: "release schedule", limit: 5 });
  for (const item of remembered) console.log(`- ${String(item.value.content)}`);
}

await main();
